src/script/miniscript.h

```diff
@@ -0,0 +1,998 @@
+ *   - This generally requires 'e' for all subexpressions which are invoked for that
+ *     dissatifsaction, and property 'f' for the unexecuted subexpressions in that case.
+ *   - Conflicts with type 'V'.
+ * - "f" Forced:
+ *   - Dissatisfactions (if any) for this expression always involve at least one signature.
+ *   - Is always true for type 'V'.
+ * - "s" Safe:
+ *   - Satisfactions for this expression always involve at least one signature.
+ * - "m" Nonmalleable:
+ *   - For every way this expression can be satisfied (which may be none),
+ *     a nonmalleable satisfaction exists.
+ *   - This generally requires 'm' for all subexpressions, and 'e' for all subexpressions
+ *     which are dissatisfied when satisfying the parent.
+ *
+ * One type property is an implementation detail:
+ * - "x" Expensive verify:
+ *   - Expressions with this property have a script whose last opcode is not EQUAL, CHECKSIG, or CHECKMULTISIG.
+ *   - Not having this property means that it can be converted to a V at no cost (by switching to the
+ *     -VERIFY version of the last opcode).
+ *
+ * Five more type properties for representing timelock information. Spend paths
+ * in miniscripts containing conflicting timelocks and heightlocks cannot be spent together.
+ * This helps users detect if miniscript does not match the semantic behaviour the
+ * user expects.
+ * - "g" Whether the branch contains a relative time timelock
+ * - "h" Whether the branch contains a relative height timelock
+ * - "i" Whether the branch contains a absolute time timelock
+ * - "j" Whether the branch contains a absolute time heightlock
+ * - "k"
+ *   - Whether all satisfactions of this expression don't contain a mix of heightlock and timelock
+ *     of the same type.
+ *   - If the miniscript does not have the "k" property, the miniscript template will not match
+ *     the user expectation of the corresponding spending policy.
+ * For each of these properties the subset rule holds: an expression with properties X, Y, and Z, is also
+ * valid in places where an X, a Y, a Z, an XY, ... is expected.
+*/
+class Type {
+    //! Internal bitmap of properties (see ""_mst operator for details).
+    uint32_t m_flags;
+
+    //! Internal constructor used by the ""_mst operator.
+    explicit constexpr Type(uint32_t flags) : m_flags(flags) {}
+
+public:
+    //! The only way to publicly construct a Type is using this literal operator.
+    friend constexpr Type operator"" _mst(const char* c, size_t l);
+
+    //! Compute the type with the union of properties.
+    constexpr Type operator|(Type x) const { return Type(m_flags | x.m_flags); }
+
+    //! Compute the type with the intersection of properties.
+    constexpr Type operator&(Type x) const { return Type(m_flags & x.m_flags); }
+
+    //! Check whether the left hand's properties are superset of the right's (= left is a subtype of right).
+    constexpr bool operator<<(Type x) const { return (x.m_flags & ~m_flags) == 0; }
```

Comment:
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

`<<` confused me for a while as I am used to it meaning bit shift when used in a flags like context, rather than a containment operator.